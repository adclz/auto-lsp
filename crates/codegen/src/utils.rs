pub(crate) fn sanitize_string(string: &str) -> String {
    if let Some(v) = PUNCTUATION.get(string) {
        v.to_string()
    } else if let Some(v) = RUST_KEYWORDS.get(string) {
        v.to_string()
    } else {
        let mut result = String::new();
        for c in string.chars() {
            if c == '_' {
                result.push(c);
            } else if let Some(v) = PUNCTUATION.get(&c.to_string()) {
                result.push_str(v);
            } else {
                result.push(c);
            }
        }
        result.replace(" ", "_")
    }
}

pub(crate) fn sanitize_string_to_pascal(string: &str) -> String {
    if let Some(v) = PUNCTUATION.get(string) {
        v.to_string()
    } else if let Some(v) = RUST_KEYWORDS.get(string) {
        v.to_string()
    } else {
        let mut result = String::new();
        for c in string.chars() {
            if c == '_' {
                result.push(c);
            } else if let Some(v) = PUNCTUATION.get(&c.to_string()) {
                result.push_str(v);
            } else {
                result.push(c);
            }
        }
        result
            .replace(" ", "_")
            .split('_')
            .map(|word| {
                let mut chars = word.chars();
                match chars.next() {
                    Some(first) => format!(
                        "{}{}",
                        first.to_ascii_uppercase(),
                        chars.collect::<String>()
                    ),
                    None => String::new(),
                }
            })
            .collect()
    }
}

pub static PUNCTUATION: phf::Map<&'static str, &'static str> = phf::phf_map! {
    "{" => "LeftCurly",
    "}" => "RightCurly",
    "(" => "LeftParen",
    ")" => "RightParen",
    "[" => "LeftBracket",
    "]" => "RightBracket",
    "," => "Comma",
    ":" => "Colon",
    ";" => "Semicolon",
    "." => "Dot",
    "@" => "At",
    "!" => "Bang",
    "#" => "Hash",
    "$" => "Dollar",
    "%" => "Percent",
    "^" => "Caret",
    "&" => "Ampersand",
    "*" => "Star",
    "-" => "Minus",
    "_" => "Underscore",
    "+" => "Plus",
    "=" => "Equal",
    ">" => "Greater",
    "<" => "Less",
    "|" => "Pipe",
    "~" => "Tilde",
    "/" => "Slash",
    "\\" => "Backslash",
    "//" => "SlashSlash",
    "//=>" => "SlashSlashEqual",
    "//=" => "SlashSlashEqual",
    "/=" => "SlashEqual",
    "/>" => "SlashGreater",
    "/?" => "SlashQuestion",
    "/??" => "SlashNullish",
    "/*" => "SlashStar",
    "*/" => "StarSlash",
    "+++" => "PlusPlusPlus",
    "!!" => "BangBang",
    "!!=" => "BangBangEqual",
    "!!?" => "BangBangQuestion",
    "!!??" => "BangBangNullish",
    "!!???" => "BangBangNullishQuestion",
    "?" => "Question",
    "->" => "Arrow",
    "=>" => "FatArrow",
    "++" => "PlusPlus",
    "--" => "MinusMinus",
    "&&" => "And",
    "||" => "Or",
    "==" => "EqualEqual",
    "!=" => "NotEqual",
    ">=" => "GreaterEqual",
    "<=" => "LessEqual",
    "===" => "TripleEqual",
    "!==" => "NotTripleEqual",
    "<<" => "ShiftLeft",
    ">>" => "ShiftRight",
    ">>>" => "ShiftRightUnsigned",
    "+=" => "PlusEqual",
    "-=" => "MinusEqual",
    "*=" => "StarEqual",
    "%=" => "PercentEqual",
    "&=" => "AmpersandEqual",
    "|=" => "PipeEqual",
    "^=" => "CaretEqual",
    "&&=" => "AndEqual",
    "||=" => "OrEqual",
    "??=" => "NullishEqual",
    "??" => "Nullish",
    "???" => "NullishQuestion",
    "**" => "StarStar",
    "**=" => "StarStarEqual",
    "<>" => "LessGreater",
    "<=>" => "LessGreaterEqual",
};

pub static RUST_KEYWORDS: phf::Map<&'static str, &'static str> = phf::phf_map! {
    "abstract" => "Abstract",
    "as" => "As",
    "async" => "Async",
    "await" => "Await",
    "break" => "Break",
    "const" => "Const",
    "continue" => "Continue",
    "crate" => "Crate",
    "dyn" => "Dyn",
    "else" => "Else",
    "enum" => "Enum",
    "extern" => "Extern",
    "false" => "False",
    "final" => "Final",
    "fn" => "Fn",
    "for" => "For",
    "if" => "If",
    "impl" => "Impl",
    "in" => "In",
    "let" => "Let",
    "loop" => "Loop",
    "match" => "Match",
    "mod" => "Mod",
    "move" => "Move",
    "mut" => "Mut",
    "pub" => "Pub",
    "ref" => "Ref",
    "return" => "Return",
    "self" => "Self",
    "static" => "Static",
    "struct" => "Struct",
    "super" => "Super",
    "trait" => "Trait",
    "true" => "True",
    "type" => "Type",
    "unsafe" => "Unsafe",
    "use" => "Use",
    "where" => "Where",
    "while" => "While",
    "with" => "With",
    "yield" => "Yield",
    "None" => "_None",
    "Some" => "_Some",
    "Ok" => "_Ok",
    "Err" => "_Err",
    "Result" => "_Result",
    "Option" => "_Option",
    "Vec" => "_Vec",
};
