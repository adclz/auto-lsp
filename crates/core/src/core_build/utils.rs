/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

pub(crate) fn intersecting_ranges(
    range1: &std::ops::Range<usize>,
    range2: &tree_sitter::Range,
) -> bool {
    range1.start <= range2.start_byte && range1.end >= range2.end_byte
}

pub(crate) fn tree_sitter_range_to_lsp_range(range: &tree_sitter::Range) -> lsp_types::Range {
    let start = range.start_point;
    let end = range.end_point;
    lsp_types::Range {
        start: lsp_types::Position {
            line: start.row as u32,
            character: start.column as u32,
        },
        end: lsp_types::Position {
            line: end.row as u32,
            character: end.column as u32,
        },
    }
}
