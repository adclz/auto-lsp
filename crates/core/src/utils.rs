#[macro_export]
macro_rules! dispatch_once {
    ($node:expr, [$($ty:ty => $method:ident($($param:expr),*)),*]) => {
        $(
            if let Some(n) = $node.downcast_ref::<$ty>() {
                return n.$method($($param),*);
            };
        )*
    };
}

#[macro_export]
macro_rules! dispatch {
    ($node:expr, [$($ty:ty => $method:ident($($param:expr),*)),*]) => {
        $(
            if let Some(n) = $node.downcast_ref::<$ty>() {
                n.$method($($param),*)?;
            };
        )*
    };
}