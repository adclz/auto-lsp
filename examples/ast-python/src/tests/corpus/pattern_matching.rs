/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

use super::utils::Result;
use crate::snap;

#[test]
fn matching_specific_values() -> Result {
    snap!(
        r#"
match command.split():
    case ["quit"]:
        print("Goodbye!")
        quit_game()
    case ["look"]:
        current_room.describe()
    case ["get", obj]:
        character.get(obj, current_room)
    case ["go", direction]:
        current_room = current_room.neighbor(direction)
    # The rest of your commands go here
"#
    )
}

#[test]
fn matching_multiple_values() -> Result {
    snap!(
        r#"
match command.split():
    case ["drop", *objects]:
        for obj in objects:
            character.drop(obj, current_room)
"#
    )
}

#[test]
fn adding_a_wild_card() -> Result {
    snap!(
        r#"
match command.split():
# ^ conditional
    case ["quit"]: ... # Code omitted for brevity
    case ["go", direction]: pass
    case ["drop", *objects]: pass
    case _:
        print(f"Sorry, I couldn't understand {command!r}")
"#
    )
}

#[test]
fn or_patterns() -> Result {
    snap!(
        r#"
match command.split():
    case ["north"] | ["go", "north"]:
        current_room = current_room.neighbor("north")
    case ["get", obj] | ["pick", "up", obj] | ["pick", obj, "up"]:
        pass
"#
    )
}

#[test]
fn as_patterns() -> Result {
    snap!(
        r#"
match command.split():
    case ["go", ("north" | "south" | "east" | "west") as direction]:
        current_room = current_room.neighbor(direction)
"#
    )
}

#[test]
fn actually_not_match() -> Result {
    snap!(
        r#"
match = 2
match, a = 2, 3
match: int = secret
x, match = 2, "hey, what's up?"
*match, last = [1, 2, 3]
def foo(**match): pass
"#
    )
}

#[test]
fn match_is_match_but_not_pattern_matching() -> Result {
    snap!(
        r#"
a = [match]
match = [match]
"#
    )
}

#[test]
fn match_kwargs() -> Result {
    snap!(
        r#"
field = call(match=r".*\.txt$")
"#
    )
}

#[test]
fn match_kwargs_2() -> Result {
    snap!(
        r#"
field = match(match=match, match)
"#
    )
}

#[test]
fn case_used_as_identifier() -> Result {
    snap!(
        r#"
a = [case]
case = [case]
just_in_case = call_me(case=True)
"#
    )
}

#[test]
fn if_guards() -> Result {
    snap!(
        r#"
match 0:
    case 0 if False:
        x = False
    case 0 if True:
        x = True
"#
    )
}

#[test]
fn literals() -> Result {
    snap!(
        r#"
match xxx:
    case 3 | -3:
      pass
    case "something":
      pass
    case "something" "else":
      pass
    case 1.0 | -1.0:
      pass
    case True | False:
      pass
    case None:
      pass
"#
    )
}

#[test]
fn comma_separated_cases() -> Result {
    snap!(
        r#"
match (0, 1, 2):
    case 0,1:
        x = 0
    case 0, *x:
        x = 0
"#
    )
}

#[test]
fn case_terminating_in_comma() -> Result {
    snap!(
        r#"
match x,:
    case *x,:
        y = 0
"#
    )
}

#[test]
fn multiple_match_patterns() -> Result {
    snap!(
        r#"
match ..., ...:
    case a, b:
        return locals()
"#
    )
}

#[test]
fn match_match_case_case() -> Result {
    snap!(
        r#"
match = case = 0
match match:
    case case:
        x = 0
"#
    )
}

#[test]
fn walrus_match_issue_150() -> Result {
    snap!(
        r#"
if match := re.fullmatch(r"(-)?(\d+:)?\d?\d:\d\d(\.\d*)?", time, flags=re.ASCII):
    return 42
"#
    )
}

#[test]
fn matching_objects() -> Result {
    snap!(
        r#"
match event.get():
    case Click(position=(x, y)):
        handle_click_at(x, y)
    case KeyPress(key_name="Q") | Quit():
        game.quit()
    case KeyPress(key_name="up arrow"):
        game.go_north()
        ...
    case KeyPress():
        pass # Ignore other keystrokes
    case other_event:
        raise ValueError(f"Unrecognized event: {other_event}")
"#
    )
}

#[test]
fn positional_arguments() -> Result {
    snap!(
        r#"
match event.get():
    case Click((x, y)):
        handle_click_at(x, y)
"#
    )
}

#[test]
fn constants_and_enums() -> Result {
    snap!(
        r#"
match event.get():
    case Click((x, y), button=Button.LEFT):  # This is a left click
        handle_click_at(x, y)
    case Click():
        pass  # ignore other clicks
"#
    )
}
