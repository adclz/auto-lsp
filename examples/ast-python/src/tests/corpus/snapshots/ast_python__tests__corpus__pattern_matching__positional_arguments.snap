---
source: src/tests/corpus/pattern_matching.rs
expression: module
---
Module {
    children: [
        CompoundStatement(
            MatchStatement(
                MatchStatement {
                    body: Block {
                        alternative: [
                            CaseClause {
                                guard: None,
                                consequence: Block {
                                    alternative: [],
                                    children: [
                                        SimpleStatement(
                                            ExpressionStatement(
                                                ExpressionStatement {
                                                    children: [
                                                        Expression(
                                                            PrimaryExpression(
                                                                Call(
                                                                    Call {
                                                                        function: Identifier(
                                                                            Identifier {
                                                                                [RANGE]
                                                                            },
                                                                        ),
                                                                        arguments: ArgumentList(
                                                                            ArgumentList {
                                                                                children: [
                                                                                    Expression(
                                                                                        PrimaryExpression(
                                                                                            Identifier(
                                                                                                Identifier {
                                                                                                    [RANGE]
                                                                                                },
                                                                                            ),
                                                                                        ),
                                                                                    ),
                                                                                    Expression(
                                                                                        PrimaryExpression(
                                                                                            Identifier(
                                                                                                Identifier {
                                                                                                    [RANGE]
                                                                                                },
                                                                                            ),
                                                                                        ),
                                                                                    ),
                                                                                ],
                                                                                [RANGE]
                                                                            },
                                                                        ),
                                                                        [RANGE]
                                                                    },
                                                                ),
                                                            ),
                                                        ),
                                                    ],
                                                    [RANGE]
                                                },
                                            ),
                                        ),
                                    ],
                                    [RANGE]
                                },
                                children: [
                                    CasePattern {
                                        children: Some(
                                            ClassPattern(
                                                ClassPattern {
                                                    children: [
                                                        DottedName(
                                                            DottedName {
                                                                children: [
                                                                    Identifier {
                                                                        [RANGE]
                                                                    },
                                                                ],
                                                                [RANGE]
                                                            },
                                                        ),
                                                        CasePattern(
                                                            CasePattern {
                                                                children: Some(
                                                                    TuplePattern(
                                                                        TuplePattern {
                                                                            children: [
                                                                                CasePattern(
                                                                                    CasePattern {
                                                                                        children: Some(
                                                                                            DottedName(
                                                                                                DottedName {
                                                                                                    children: [
                                                                                                        Identifier {
                                                                                                            [RANGE]
                                                                                                        },
                                                                                                    ],
                                                                                                    [RANGE]
                                                                                                },
                                                                                            ),
                                                                                        ),
                                                                                        [RANGE]
                                                                                    },
                                                                                ),
                                                                                CasePattern(
                                                                                    CasePattern {
                                                                                        children: Some(
                                                                                            DottedName(
                                                                                                DottedName {
                                                                                                    children: [
                                                                                                        Identifier {
                                                                                                            [RANGE]
                                                                                                        },
                                                                                                    ],
                                                                                                    [RANGE]
                                                                                                },
                                                                                            ),
                                                                                        ),
                                                                                        [RANGE]
                                                                                    },
                                                                                ),
                                                                            ],
                                                                            [RANGE]
                                                                        },
                                                                    ),
                                                                ),
                                                                [RANGE]
                                                            },
                                                        ),
                                                    ],
                                                    [RANGE]
                                                },
                                            ),
                                        ),
                                        [RANGE]
                                    },
                                ],
                                [RANGE]
                            },
                        ],
                        children: [],
                        [RANGE]
                    },
                    subject: [
                        PrimaryExpression(
                            Call(
                                Call {
                                    function: Attribute(
                                        Attribute {
                                            attribute: Identifier {
                                                [RANGE]
                                            },
                                            object: Identifier(
                                                Identifier {
                                                    [RANGE]
                                                },
                                            ),
                                            [RANGE]
                                        },
                                    ),
                                    arguments: ArgumentList(
                                        ArgumentList {
                                            children: [],
                                            [RANGE]
                                        },
                                    ),
                                    [RANGE]
                                },
                            ),
                        ),
                    ],
                    [RANGE]
                },
            ),
        ),
    ],
    [RANGE]
}
